### DISCUSSION TITLE
AnyBlock的实时编辑体验

### DISCUSSION THREAD

**Discussion by LincZero** [any-block/any-block#80] [General]
## 计划

一步到位太麻烦了，而且一做，可能标签页、卡片、表格单元格等都要全部单独适配。思来想去，先加了个简化版本的todo

（仅todo，简化版本也得有空再做，优先级不高，我目前优先级高的还是主要在网页版本（不需要编辑交互）的扩展中）

## Step1 新的统一编辑窗口

提供一个新的编辑按钮：无论是什么类型的anyblock，统一给出anyblock内所有内容源md编辑区（不区分位于哪一个单元格/卡片/标签页，全部一起编辑）

![image](https://github.com/user-attachments/assets/37464990-9aac-48ea-bde5-24dcb1c91ba3)

## Step2 单独适配

后续如果有人再反映……就再单独做 "标签页、卡片、表格单元格、脑图节点图" 等的独立适配

只能一个一个做……如果只做了一个标签页……肯定又有人要求再支持其他ab块了。工作量挺大的。到时有空再反馈什么做什么吧

**Comment by playermiller109** [comment]
用额外弹窗实现编辑不是个办法。我写过类似原理的插件 [小插件：重命名小标题时更新内部链接防止断裂](https://forum-zh.obsidian.md/t/topic/38592/1)，~~现在 v1.7.4 出可编辑浮窗鼠标挪过去掉 hover~~ 更新：现已修好。话题里面提的 obsidian-tabs 插件我刚试了也是一样的。

<img width="300" src="https://github.com/user-attachments/assets/284a32f4-5a93-42a9-9ef1-9cdbb21965db">

目前我所知唯一看表现可能可行的编辑方式是在 v1.5.0 版本更新前的一个社区插件 [ob-table-enhancer](https://github.com/Stardusten/ob-table-enhancer)，我给论坛另外一个人介绍过。

<img width="500" src="https://github.com/user-attachments/assets/d458ad3d-9d48-4f7a-9b6c-2d2d59457e27">

**Comment by LincZero** [comment]
不弹窗，当前的设计是：直接AB块的渲染内容隐藏起来，替换为文本输入框。
我想做Typora的引用块那种体验。而这种设计也是和Typora的引用块编辑特别是嵌套引用块的编辑的相同的。

实时编辑目前是两种方案：

- 一是obsidain大部分场景中**直接**使用codeMirror来渲染，并且是ob实时模式中每一行都是一个 `<div class="HyperMd-xxx cm-line">`
  - 应用：包括ob的列表、引用块
  - 优点：简单，交互的部分使用现成的CM引擎，无需额外操作和适配
  - 缺点：不存在嵌套，每一行都只是一个  `<div class="HyperMd-xxx cm-line">`，其 “嵌套” 是通过设置该div的缩进实现的。引用块里面嵌列表、代码块等都是失效的
- 二是不使用CodeMirror，而是渲染一个 “可交互” 的HTML元素，该HTML元素中允许包括输入框部分（也可以是HTML中再内嵌CM部分），并能在交互时自动修改源文本。
  - 应用：Typora的列表、引用块、表格，以及Ob的新表格和之前的table-enhance插件等，都属于这种方式。
  - 缺点：是得一一适配HTML内 **局部输入框与源文本的某一个位置** 的对应关系，非常麻烦。而且需要特别注意不能错误操作源文本，导致文本丢失
  - 优点：交互性特别好，并且对于 “嵌套” 情况的适配性更好，是确实有html的嵌套存在。另一个优点是能抄作业……有个 table-enhance 插件可以参考

这里也采用方案二（真想用方案二把ob的实时模式给重写了……）

**Comment by LincZero** [comment]
发现刚刚说错了，其实这两不冲突，HTML元素里面可以嵌套CodeMirror。

只是ob他是每一行的文本对应一个div，然后对应一个codemirror编辑区，这样的话编辑CodeMirror区域后的变化就完全反映到源md中的一行里。
即：源md一行 -> 对应一个可编辑的div -> 对应一个codemirror区

而HTML显示的局部（例如单元格）也可以对应一个codemirror编辑区，只是div不是对应于一行的，而是对应于一个单元格的即源md中的一段文本。
即md被拆分成很多个部分(不一定是一行，可能是一行中的某几个字) -> 每个部分对应一个可编辑的div -> 也对应一个codemirror区 (也可以是普通的文本输入框)
而这里的拆分需要手动适配

---

你图中的这个问题，应该还是弹窗的问题吧。可能需要监听点击事件？然后如果你的弹窗存在，就禁止默认的点击行为。除非监听到单机事件中点击的内容是你弹窗中的确认或取消或监听Enter键，才自行处理

（不过看起来点击穿透了？z-index的问题？先看一下监听点击事件时，能不能抓到点击的对象是弹窗）

**Reply by playermiller109** [reply]
可能是 CM 焦点问题， Obsidian 新版浮窗可能是判断鼠标是否悬停在浮窗区域和现在取得焦点是哪个 CM 之类来判断 hover。

**Reply by LincZero** [reply]
不对啊，tabs的浮窗没出现点击穿透吧，他是点浮窗外面才消失，这也是预期行为啊。你gif图里的直接点击穿透了

不过他那个不是用ob的CM好像

**Reply by playermiller109** [reply]
我的锅，我以为 Obsidian 新版弹窗对自己的功能做了浮窗适配，所以有一个多条件的综合判断，结果它右键菜单点了浮窗也会立马消失，太捞了，看样子就是统一的问题。

**Reply by playermiller109** [reply]
<img width="500" src="https://github.com/user-attachments/assets/cd501aa2-92bb-4593-8ffe-78bfde35a7d0">

找到了问题所在的代码。给我插件生成的元素加上 `evt.stopPropagation()` 就搞定了。谢谢你啦。

**Comment by LincZero** [comment]
目前依旧还未动工，但可行性在我的另一项目中（偏实验）取得了不错的进展：

https://github.com/LincZero/obsidian-shiki-plugin

https://github.com/user-attachments/assets/701987af-753b-4fa8-a4de-7f9f9ecfa9f1

可以继续观望

**Reply by LincZero** [reply]
一个追踪进度、且允许在线尝试的 Demo: (切换到 `MdCodeMirror` 标签页使用)

https://linczero.github.io/obsidian-node-flow/EditableCodeblock.html

**Comment by LincZero** [comment]
请关注 #189
